include/arch/arm/aarch64/lib_helpers.h

```diff
@@ -0,0 +1,160 @@
+({									\
+	__asm__ volatile ("msr " STRINGIFY(reg) ", xzr"			\
+			  ::: "memory");				\
+})
+
+#define MAKE_REG_HELPER(reg)						\
+	static ALWAYS_INLINE uint64_t read_##reg(void)			\
+	{								\
+		return read_sysreg(reg);				\
+	}								\
+	static ALWAYS_INLINE void write_##reg(uint64_t val)		\
+	{								\
+		write_sysreg(val, reg);					\
+	}								\
+	static ALWAYS_INLINE void zero_##reg(void)			\
+	{								\
+		zero_sysreg(reg);					\
+	}
+
+#define MAKE_REG_HELPER_EL123(reg) \
+	MAKE_REG_HELPER(reg##_el1) \
+	MAKE_REG_HELPER(reg##_el2) \
+	MAKE_REG_HELPER(reg##_el3)
+
+MAKE_REG_HELPER(cntfrq_el0);
+MAKE_REG_HELPER(cnthctl_el2);
+MAKE_REG_HELPER(cnthp_ctl_el2);
+MAKE_REG_HELPER(cntv_ctl_el0)
+MAKE_REG_HELPER(cntv_cval_el0)
+MAKE_REG_HELPER(cntvct_el0);
+MAKE_REG_HELPER(cntvoff_el2);
+MAKE_REG_HELPER(currentel);
+MAKE_REG_HELPER(daif)
+MAKE_REG_HELPER(hcr_el2);
+MAKE_REG_HELPER(id_aa64pfr0_el1);
+MAKE_REG_HELPER(scr_el3);
+
+MAKE_REG_HELPER_EL123(actlr)
+MAKE_REG_HELPER_EL123(cpacr)
+MAKE_REG_HELPER_EL123(cptr)
+MAKE_REG_HELPER_EL123(elr)
+MAKE_REG_HELPER_EL123(esr)
+MAKE_REG_HELPER_EL123(far)
+MAKE_REG_HELPER_EL123(mair)
+MAKE_REG_HELPER_EL123(sctlr)
+MAKE_REG_HELPER_EL123(spsr)
+MAKE_REG_HELPER_EL123(tcr)
+MAKE_REG_HELPER_EL123(ttbr0)
+MAKE_REG_HELPER_EL123(vbar)
+
+static ALWAYS_INLINE void enable_debug_exceptions(void)
+{
+	__asm__ volatile ("msr DAIFClr, %0"
+			  :: "i" (DAIFCLR_DBG_BIT) : "memory");
+}
+
+static ALWAYS_INLINE void disable_debug_exceptions(void)
+{
+	__asm__ volatile ("msr DAIFSet, %0"
+			  :: "i" (DAIFSET_DBG_BIT) : "memory");
+}
+
+static ALWAYS_INLINE void enable_serror_exceptions(void)
+{
+	__asm__ volatile ("msr DAIFClr, %0"
+			  :: "i" (DAIFCLR_ABT_BIT) : "memory");
+}
+
+static ALWAYS_INLINE void disable_serror_exceptions(void)
+{
+	__asm__ volatile ("msr DAIFSet, %0"
+			  :: "i" (DAIFSET_ABT_BIT) : "memory");
+}
+
+static ALWAYS_INLINE void enable_irq(void)
+{
+	__asm__ volatile ("msr DAIFClr, %0"
+			  :: "i" (DAIFCLR_IRQ_BIT) : "memory");
+}
+
+static ALWAYS_INLINE void disable_irq(void)
+{
+	__asm__ volatile ("msr DAIFSet, %0"
+			  :: "i" (DAIFSET_IRQ_BIT) : "memory");
+}
+
+static ALWAYS_INLINE void enable_fiq(void)
+{
+	__asm__ volatile ("msr DAIFClr, %0"
+			  :: "i" (DAIFCLR_FIQ_BIT) : "memory");
+}
+
+static ALWAYS_INLINE void disable_fiq(void)
+{
+	__asm__ volatile ("msr DAIFSet, %0"
+			  :: "i" (DAIFSET_FIQ_BIT) : "memory");
+}
+
+#define dsb()	__asm__ volatile ("dsb sy" ::: "memory")
+#define dmb()	__asm__ volatile ("dmb sy" ::: "memory")
+#define isb()	__asm__ volatile ("isb" ::: "memory")
+
+/* Zephyr needs these as well */
+#define __ISB() isb()
+#define __DMB() dmb()
+#define __DSB() dsb()
+
+static inline bool is_el_implemented(unsigned int el)
+{
+	unsigned int shift;
+
+	if (el > 3) {
+		return false;
+	}
+
+	shift = ID_AA64PFR0_EL1_SHIFT * el;
+
+	return (((read_id_aa64pfr0_el1() >> shift) & ID_AA64PFR0_ELX_MASK) != 0U);
+}
+
+static inline bool is_el2_sec_supported(void)
```

Comment:
...Isn't this identical to `is_el_implemented(2)`?

Reply:
nope, here we are checking the `SEL2` field of `ID_AA64PFR0_EL1`.

Reply:
...oh, I saw that, saw the `EL2` field, and assumed it was looking at `EL2` not `SEL2`.

I assume `SEL2` was added by the `ARMv8.4-SecEL2` extension?

Is there a public document that I could take a look at for `ARMv8.4-SecEL2` akin to the "ARM Architecture Reference Manual Supplement ARMv8.1, for ARMv8-A architecture profile"?

Reply:
>akin to the "ARM Architecture Reference Manual Supplement ARMv8.1, for ARMv8-A architecture profile"?

Nevermind, ARM's latest Architecture Reference Manual includes the ARMv8.x supplements.

Reply:
`SEL2` != `EL2`, right.